src/imitation/algorithms/tabular_irl.py

```diff
@@ -171,294 +177,94 @@ def mce_irl(
     # number of optimisation steps taken
     t = 0
     assert demo_state_om.shape == (len(obs_mat),)
-    opt_init, opt_update, opt_get_params = optimiser_tuple
-    rew_params = rmodel.get_params()
-    opt_state = opt_init(rew_params)
 
     while linf_delta > linf_eps and grad_norm > grad_l2_eps:
+        # zero grad
+        optimizer.zero_grad()
+
         # get reward predicted for each state by current model, & compute
         # expected # of times each state is visited by soft-optimal policy
         # w.r.t that reward function
-        predicted_r, out_grads = rmodel.out_grads(obs_mat)
-        _, visitations = mce_occupancy_measures(env, R=predicted_r)
-        # gradient of partition function w.r.t parameters; equiv to expectation
-        # over states drawn from current imitation distribution of the gradient
-        # of the reward function w.r.t its params
-        pol_grad = np.sum(visitations[:, None] * out_grads, axis=0)
-        # gradient of reward function w.r.t parameters, with expectation taken
-        # over states
-        expert_grad = np.sum(demo_state_om[:, None] * out_grads, axis=0)
-        grad = pol_grad - expert_grad
+        predicted_r = squeeze_r(rmodel(torch_obs_mat))
+        assert predicted_r.shape == (obs_mat.shape[0],)
+        predicted_r_np = predicted_r.detach().cpu().numpy()
+        _, visitations = mce_occupancy_measures(env, R=predicted_r_np)
+
+        # Forward/back/step (grads are zeroed at the top).
```

Comment:
I found this a little hard to follow. Given that the "loss" is unusual, can you be a bit more explicit? e.g. write:
```python
# Forward/back/step (grads are zeroed at the top).
# weights_th(s) = \pi(s) - D(s)
weights_th = th.as_tensor(
    visitations - demo_state_om, dtype=dtype, device=device
)
# The "loss" is then:
#   E_\pi[r_\theta(S)] - E_D[r_\theta(S)]
loss = th.dot(weights_th, predicted_r)
# This gives the required gradient:
#   E_\pi[\nabla r_\theta(S)] - E_D[\nabla r_\theta(S)].
loss.backward()
```